File: .github/update_nasdaq_snapshot.py
import os
import time
import json
import traceback
from datetime import datetime, timezone
from urllib import request, error

try:
    from dotenv import load_dotenv
    load_dotenv()
except Exception:
    pass

SUPABASE_URL = os.getenv("SUPABASE_URL")
SUPABASE_SERVICE_ROLE = os.getenv("SUPABASE_SERVICE_ROLE") or os.getenv("SUPABASE_SERVICE_ROLE_KEY")

YAHOO_QUOTE_ENDPOINT = "https://query1.finance.yahoo.com/v7/finance/quote?symbols=%5EIXIC"
YAHOO_NEWS_ENDPOINT = "https://query1.finance.yahoo.com/v2/finance/news?category=generalnews&symbol=%5EIXIC&region=US"
YAHOO_SECTOR_ENDPOINT = "https://query1.finance.yahoo.com/v1/finance/screener/predefined/saved?scrIds=sector_etf&count=25"
YAHOO_HEATMAP_ENDPOINT = "https://query1.finance.yahoo.com/v7/finance/quote?symbols={symbols}"
NASDAQ_ADVANCERS_ENDPOINT = "https://api.nasdaq.com/api/marketmovers?type=advancers&exchange=nasdaq"
NASDAQ_DECLINERS_ENDPOINT = "https://api.nasdaq.com/api/marketmovers?type=decliners&exchange=nasdaq"

BIG_TECH_SYMBOLS = ["AAPL", "MSFT", "NVDA", "GOOGL", "AMZN", "META", "TSLA", "AVGO", "NFLX", "ADBE"]

NASDAQ_HEADERS = {
    "Accept": "application/json, text/plain, */*",
    "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, مثل Gecko) Chrome/124.0.0.0 Safari/537.36",
    "Origin": "https://www.nasdaq.com",
    "Referer": "https://www.nasdaq.com/",
}

YAHOO_HEADERS = {
    "Accept": "application/json, text/plain, */*",
    "User-Agent": NASDAQ_HEADERS["User-Agent"],
    "Connection": "keep-alive",
    "Accept-Language": "en-US,en;q=0.9",
}


def log(message: str):
    print(f"[update_nasdaq] {message}")


def fetch_json(url: str, headers: dict | None = None, max_retries: int = 4, backoff: float = 1.5):
    attempt = 0
    hdrs = headers or {}
    if not hdrs:
        hdrs = YAHOO_HEADERS
    else:
        # ensure user agent exists to avoid 429s
        hdrs = {**YAHOO_HEADERS, **headers}

    while True:
        try:
            req = request.Request(url, headers=hdrs)
            with request.urlopen(req, timeout=30) as resp:
                raw = resp.read().decode("utf-8")
                return json.loads(raw)
        except error.HTTPError as http_err:
            status = getattr(http_err, 'code', None)
            if status in (429, 503) and attempt < max_retries:
                delay = backoff ** attempt
                log(f"HTTP {status} from {url}; retrying in {delay:.1f}s")
                time.sleep(delay)
                attempt += 1
                continue
            raise
        except error.URLError as url_err:
            if attempt < max_retries:
                delay = backoff ** attempt
                log(f"URL error {url_err.reason} for {url}; retrying in {delay:.1f}s")
                time.sleep(delay)
                attempt += 1
                continue
            raise


def parse_number(value):
    if value is None:
        return None
    if isinstance(value, (int, float)):
        return float(value)
    try:
        cleaned = str(value).replace(",", "").strip()
        if not cleaned:
            return None
        return float(cleaned)
    except Exception:
        return None


def fetch_quote():
    data = fetch_json(YAHOO_QUOTE_ENDPOINT)
    quote = (data.get("quoteResponse", {}).get("result") or [None])[0]
    if not quote:
        raise RuntimeError("Yahoo Finance quote response empty")
    ts = quote.get("regularMarketTime")
    trading_dt = datetime.fromtimestamp(ts, timezone.utc) if ts else datetime.now(timezone.utc)
    return {
        "trading_date": trading_dt.date().isoformat(),
        "close": parse_number(quote.get("regularMarketPrice", quote.get("regularMarketPreviousClose"))),
        "change": parse_number(quote.get("regularMarketChange")),
        "change_percent": parse_number(quote.get("regularMarketChangePercent")),
        "open": parse_number(quote.get("regularMarketOpen")),
        "high": parse_number(quote.get("regularMarketDayHigh")),
        "low": parse_number(quote.get("regularMarketDayLow")),
        "volume": parse_number(quote.get("regularMarketVolume")),
    }


def fetch_advancers_decliners():
    try:
        adv_data = fetch_json(NASDAQ_ADVANCERS_ENDPOINT, NASDAQ_HEADERS)
        dec_data = fetch_json(NASDAQ_DECLINERS_ENDPOINT, NASDAQ_HEADERS)
        return {
            "advancers": parse_number(adv_data.get("data", {}).get("totalRecords")) or parse_number(adv_data.get("data", {}).get("count")) or None,
            "decliners": parse_number(dec_data.get("data", {}).get("totalRecords")) or parse_number(dec_data.get("data", {}).get("count")) or None,
        }
    except Exception as exc:
        log(f"warn: advancers/decliners unavailable -> {exc}")
        return {"advancers": None, "decliners": None}


def fetch_sector_performance():
    try:
        data = fetch_json(YAHOO_SECTOR_ENDPOINT, {"User-Agent": NASDAQ_HEADERS["User-Agent"]})
        quotes = data.get("finance", {}).get("result", [{}])[0].get("quotes", [])
        sectors = []
        for item in quotes:
            perf = parse_number(item.get("regularMarketChangePercent") or item.get("ytdReturn"))
            sectors.append({
                "symbol": item.get("symbol"),
                "name": item.get("shortName") or item.get("longName"),
                "change_percent": perf,
            })
        ranked = [s for s in sectors if s.get("change_percent") is not None]
        ranked.sort(key=lambda r: r["change_percent"], reverse=True)
        leading = ranked[0]["name"] if ranked else None
        lagging = ranked[-1]["name"] if ranked else None
        return sectors, leading, lagging
    except Exception as exc:
        log(f"warn: sector performance unavailable -> {exc}")
        return [], None, None


def fetch_heatmap():
    try:
        symbols_param = "%2C".join(BIG_TECH_SYMBOLS)
        endpoint = YAHOO_HEATMAP_ENDPOINT.format(symbols=symbols_param)
        data = fetch_json(endpoint)
        items = data.get("quoteResponse", {}).get("result", [])
        heatmap = []
        for item in items:
            heatmap.append({
                "symbol": item.get("symbol"),
                "name": item.get("longName") or item.get("shortName"),
                "change_percent": parse_number(item.get("regularMarketChangePercent")),
                "market_cap": parse_number(item.get("marketCap")),
            })
        return heatmap
    except Exception as exc:
        log(f"warn: heatmap unavailable -> {exc}")
        return []


def fetch_headline():
    try:
        data = fetch_json(YAHOO_NEWS_ENDPOINT)
        stream = data.get("data", {}).get("main", {}).get("stream", [])
        for item in stream:
            title = item.get("title")
            link = item.get("link")
            if title and link:
                return title, link
    except Exception as exc:
        log(f"warn: headline unavailable -> {exc}")
    return None, None


def call_supabase(payload: dict):
    if not SUPABASE_URL or not SUPABASE_SERVICE_ROLE:
        raise RuntimeError("Missing SUPABASE_URL or SUPABASE_SERVICE_ROLE")

    url = f"{SUPABASE_URL}/rest/v1/rpc/upsert_nasdaq_daily_snapshot"
    body = json.dumps(payload).encode("utf-8")
    headers = {
        "Content-Type": "application/json",
        "apikey": SUPABASE_SERVICE_ROLE,
        "Authorization": f"Bearer {SUPABASE_SERVICE_ROLE}",
    }
    req = request.Request(url, data=body, headers=headers, method="POST")
    with request.urlopen(req, timeout=30) as resp:
        resp_body = resp.read().decode("utf-8")
        if resp.status >= 400:
            raise RuntimeError(f"Supabase error {resp.status}: {resp_body}")
        try:
            return json.loads(resp_body)
        except json.JSONDecodeError:
            return None


def main():
    log("Collecting Nasdaq snapshot...")
    quote = fetch_quote()
    adv_decl = fetch_advancers_decliners()
    sectors, leading, lagging = fetch_sector_performance()
    heatmap = fetch_heatmap()
    headline, headline_source = fetch_headline()

    metadata = {
        "fetched_at": datetime.now(timezone.utc).isoformat(),
        "source": "Yahoo Finance / Nasdaq API",
        "notices": [],
    }
    if adv_decl["advancers"] is None or adv_decl["decliners"] is None:
        metadata["notices"].append("advancers_decliners_unavailable")
    if not sectors:
        metadata["notices"].append("sector_performance_unavailable")
    if not heatmap:
        metadata["notices"].append("heatmap_unavailable")
    if not headline:
        metadata["notices"].append("headline_unavailable")

    payload = {
        "p_trading_date": quote["trading_date"],
        "p_close_price": quote["close"],
        "p_change_points": quote["change"],
        "p_change_percent": quote["change_percent"],
        "p_open_price": quote["open"],
        "p_high_price": quote["high"],
        "p_low_price": quote["low"],
        "p_volume": quote["volume"],
        "p_advancers_count": adv_decl["advancers"],
        "p_decliners_count": adv_decl["decliners"],
        "p_leading_sector": leading,
        "p_lagging_sector": lagging,
        "p_headline": headline,
        "p_headline_source": headline_source,
        "p_heatmap_json": heatmap,
        "p_sectors_json": sectors,
        "p_metadata_json": metadata,
    }

    log(f"Upserting snapshot for {quote['trading_date']}...")
    call_supabase(payload)
    log("Done.")


if __name__ == "__main__":
    try:
        main()
    except error.HTTPError as http_err:
        log(f"HTTP error: {http_err.code} {http_err.reason}")
        raise
    except Exception as exc:
        log(f"Failed: {exc}")
        traceback.print_exc()
        raise

File: scripts/update-nasdaq-snapshot.js
import fs from 'fs/promises';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const DEFAULT_SUPABASE_ENV_PATHS = [
  path.resolve(__dirname, '..', '.env'),
  path.resolve(__dirname, '..', 'api-server', '.env'),
  path.resolve(__dirname, '..', 'api-server', 'env'),
  path.resolve(__dirname, '..', 'env'),
];

const BIG_TECH_SYMBOLS = ['AAPL', 'MSFT', 'NVDA', 'GOOGL', 'AMZN', 'META', 'TSLA', 'AVGO', 'NFLX', 'ADBE'];
const YAHOO_QUOTE_ENDPOINT = 'https://query1.finance.yahoo.com/v7/finance/quote';
const YAHOO_NEWS_ENDPOINT = 'https://query1.finance.yahoo.com/v2/finance/news';
const YAHOO_SECTOR_ENDPOINT = 'https://query1.finance.yahoo.com/v1/finance/screener/predefined/saved?scrIds=sector_etf&count=25';
const NASDAQ_ADVANCERS_ENDPOINT = 'https://api.nasdaq.com/api/marketmovers?type=advancers&exchange=nasdaq';
const NASDAQ_DECLINERS_ENDPOINT = 'https://api.nasdaq.com/api/marketmovers?type=decliners&exchange=nasdaq';

const NASDAQ_REQUEST_HEADERS = {
  'Accept': 'application/json, text/plain, */*',
  'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
  'Origin': 'https://www.nasdaq.com',
  'Referer': 'https://www.nasdaq.com/'
};

function parseNumber(value) {
  if (value === null || typeof value === 'undefined') return null;
  if (typeof value === 'number') return value;
  const cleaned = String(value).replace(/[,\s]/g, '');
  const parsed = Number(cleaned);
  return Number.isFinite(parsed) ? parsed : null;
}

async function readSupabaseEnv() {
  const result = {
    supabaseUrl: process.env.SUPABASE_URL,
    serviceRole: process.env.SUPABASE_SERVICE_ROLE || process.env.SUPABASE_SERVICE_ROLE_KEY,
  };

  if (result.supabaseUrl && result.serviceRole) {
    return result;
  }

  for (const candidate of DEFAULT_SUPABASE_ENV_PATHS) {
    try {
      const raw = await fs.readFile(candidate, 'utf8');
      if (!result.supabaseUrl) {
        const urlMatch = raw.match(/SUPABASE_URL\s*=\s*"?([^"\n]+)"?/i);
        if (urlMatch) {
          result.supabaseUrl = urlMatch[1].trim();
        }
      }
      if (!result.serviceRole) {
        const keyMatch = raw.match(/SUPABASE_SERVICE_ROLE\s*=\s*"?([^"\n]+)"?/i);
        if (keyMatch) {
          result.serviceRole = keyMatch[1].trim();
        }
      }
      if (result.supabaseUrl && result.serviceRole) {
        break;
      }
    } catch (error) {
      // ignore missing files
    }
  }

  if (!result.supabaseUrl || !result.serviceRole) {
    throw new Error('Missing SUPABASE_URL or SUPABASE_SERVICE_ROLE. Please set them in environment variables or .env files.');
  }

  return result;
}

async function fetchJson(url, options = {}) {
  const response = await fetch(url, options);
  if (!response.ok) {
    throw new Error(`Request failed (${response.status}) ${response.statusText}`);
  }
  return response.json();
}

async function fetchNasdaqQuote() {
  const data = await fetchJson(`${YAHOO_QUOTE_ENDPOINT}?symbols=%5EIXIC`);
  const quote = data?.quoteResponse?.result?.[0];
  if (!quote) {
    throw new Error('Nasdaq quote not found from Yahoo Finance');
  }
  return {
    tradingDate: quote.regularMarketTime ? new Date(quote.regularMarketTime * 1000) : new Date(),
    close: parseNumber(quote.regularMarketPrice ?? quote.regularMarketPreviousClose),
    change: parseNumber(quote.regularMarketChange),
    changePercent: parseNumber(quote.regularMarketChangePercent),
    open: parseNumber(quote.regularMarketOpen),
    high: parseNumber(quote.regularMarketDayHigh),
    low: parseNumber(quote.regularMarketDayLow),
    volume: parseNumber(quote.regularMarketVolume),
  };
}

async function fetchAdvancersDecliners() {
  try {
    const [advancersData, declinersData] = await Promise.all([
      fetchJson(NASDAQ_ADVANCERS_ENDPOINT, { headers: NASDAQ_REQUEST_HEADERS }),
      fetchJson(NASDAQ_DECLINERS_ENDPOINT, { headers: NASDAQ_REQUEST_HEADERS }),
    ]);

    const extractTotal = (payload) => {
      const total = payload?.data?.totalRecords ?? payload?.data?.count ?? payload?.data?.rows?.length;
      return total ? Number(total) : null;
    };

    return {
      advancers: extractTotal(advancersData),
      decliners: extractTotal(declinersData),
    };
  } catch (error) {
    console.warn('⚠️  Failed to fetch advancers/decliners from Nasdaq API:', error.message);
    return { advancers: null, decliners: null };
  }
}

async function fetchSectorPerformance() {
  try {
    const data = await fetchJson(YAHOO_SECTOR_ENDPOINT, {
      headers: { 'User-Agent': NASDAQ_REQUEST_HEADERS['User-Agent'] },
    });
    const sectors = data?.finance?.result?.[0]?.quotes ?? [];
    const simplified = sectors.map((item) => ({
      symbol: item.symbol,
      shortName: item.shortName || item.longName,
      oneDayChange: parseNumber(item.regularMarketChangePercent ?? item.trailingAnnualDividendYield),
    }));

    const sorted = simplified
      .filter((item) => item.oneDayChange !== null && Number.isFinite(item.oneDayChange))
      .sort((a, b) => b.oneDayChange - a.oneDayChange);

    return {
      sectors: simplified,
      leading: sorted[0]?.shortName || null,
      lagging: sorted[sorted.length - 1]?.shortName || null,
    };
  } catch (error) {
    console.warn('⚠️  Failed to fetch sector performance:', error.message);
    return { sectors: [], leading: null, lagging: null };
  }
}

async function fetchHeatmap() {
  try {
    const symbols = BIG_TECH_SYMBOLS.join(',');
    const data = await fetchJson(`${YAHOO_QUOTE_ENDPOINT}?symbols=${encodeURIComponent(symbols)}`);
    const result = data?.quoteResponse?.result ?? [];
    return result.map((item) => ({
      symbol: item.symbol,
      longName: item.longName || item.shortName,
      weightPercent: parseNumber(item.marketCap),
      changePercent: parseNumber(item.regularMarketChangePercent),
    }));
  } catch (error) {
    console.warn('⚠️  Failed to fetch heatmap data:', error.message);
    return [];
  }
}

async function fetchHeadline() {
  try {
    const params = new URLSearchParams({
      category: 'generalnews',
      symbol: '%5EIXIC',
      region: 'US',
    });
    const data = await fetchJson(`${YAHOO_NEWS_ENDPOINT}?${params.toString()}`);
    const story = data?.data?.main?.stream?.find((item) => item.title && item.link);
    if (!story) return { headline: null, source: null };
    return {
      headline: story.title,
      source: story.link,
    };
  } catch (error) {
    console.warn('⚠️  Failed to fetch headline:', error.message);
    return { headline: null, source: null };
  }
}

function toISODate(date) {
  return date.toISOString().split('T')[0];
}

async function upsertSnapshot(supabaseUrl, serviceRole, payload) {
  const response = await fetch(`${supabaseUrl}/rest/v1/rpc/upsert_nasdaq_daily_snapshot`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      apikey: serviceRole,
      Authorization: `Bearer ${serviceRole}`,
    },
    body: JSON.stringify(payload),
  });

  if (!response.ok) {
    const text = await response.text();
    throw new Error(`Supabase RPC failed (${response.status}) ${response.statusText}: ${text}`);
  }

  return response.json().catch(() => null);
}

async function main() {
  const { supabaseUrl, serviceRole } = await readSupabaseEnv();

  console.log('📈 Fetching Nasdaq quote...');
  const quote = await fetchNasdaqQuote();

  console.log('🟩 Fetching advancers/decliners...');
  const advDecl = await fetchAdvancersDecliners();

  console.log('📊 Fetching sector performance...');
  const sectorData = await fetchSectorPerformance();

  console.log('🗞️  Fetching headline...');
  const headlineData = await fetchHeadline();

  console.log('🔥 Fetching heatmap data...');
  const heatmap = await fetchHeatmap();

  const tradingDate = toISODate(quote.tradingDate);
  const metadata = {
    fetchedAt: new Date().toISOString(),
    source: 'Yahoo Finance / Nasdaq API',
    notices: [],
  };

  if (advDecl.advancers === null || advDecl.decliners === null) {
    metadata.notices.push('advancers/decliners unavailable');
  }
  if (!sectorData.sectors.length) {
    metadata.notices.push('sector performance unavailable');
  }
  if (!heatmap.length) {
    metadata.notices.push('heatmap data unavailable');
  }

  const payload = {
    p_trading_date: tradingDate,
    p_close_price: quote.close,
    p_change_points: quote.change,
    p_change_percent: quote.changePercent,
    p_open_price: quote.open,
    p_high_price: quote.high,
    p_low_price: quote.low,
    p_volume: quote.volume,
    p_advancers_count: advDecl.advancers,
    p_decliners_count: advDecl.decliners,
    p_leading_sector: sectorData.leading,
    p_lagging_sector: sectorData.lagging,
    p_headline: headlineData.headline,
    p_headline_source: headlineData.source,
    p_heatmap_json: heatmap,
    p_sectors_json: sectorData.sectors,
    p_metadata_json: metadata,
  };

  console.log('⬆️  Upserting snapshot for', tradingDate);
  await upsertSnapshot(supabaseUrl, serviceRole, payload);
  console.log('✅ Nasdaq snapshot updated successfully.');
}

main().catch((error) => {
  console.error('❌ Failed to update Nasdaq snapshot:', error.message);
  process.exit(1);
});
